CIP-0139/README.md

```diff
@@ -0,0 +1,152 @@
+2. `lookupBSMap :: forall v . ByteString -> (() -> v) -> ByteStringMap v -> v`
+    - it returns the value for a given key in the map; if the key is missing, it returns the default value of type `() -> a` 
+    - its implementation depends on, as of the time of writing, ongoing work in supporting higher-order builtin functions; see the next section for the motivation behind this choice
+    - it uses logarithmic time in the `ByteStringMap` argument, linear time in the `ByteString` argument and constant memory
+3. `deleteBSMap :: forall v . ByteString -> ByteStringMap v -> ByteStringMap v`
+    - it returns a map without the key-value association identified by the input key
+    - it uses logarithmic time in the `ByteStringMap` argument, linear time in the `ByteString` argument and logarithmic memory in the `ByteStringMap` argument
+4. `unionBSMap :: forall v . ByteStringMap v -> ByteStringMap v -> ByteStringMap v`
+    - it merges the two input maps, left-biased when there are key collisions
+    - it uses linear time in the sum of the first and second arguments, and linear memory in the sum of the first and second arguments
+5. `bsMapToList :: forall v . ByteStringMap v -> List (Pair ByteString v)`
+    - it transforms the map into an association list
+    - the resulting list should be ordered (in ascending order of the keys)
+    - it is linear in the size of the `ByteStringMap`
+6. `listToBSMap :: forall v . List (Pair ByteString v) -> ByteStringMap v`
+    - it transforms an association list into a map, right-biased in the case of duplicate keys
+    - it has `O(n * logn)` time complexity, where `n` is the size of the list
+    - it has linear space complexity
+    - if the input list does not contain duplicate keys and is ordered (in ascending order of the keys), the space and time complexity will be linear
+
+## Rationale: how does this CIP achieve its goals?
+
+### Constructing `ByteStringMap`s
+
+The first concern we have to address is the reasoning behind not introducing a builtin function for constructing empty `ByteStringMap`s. Since maps are polymorphic in the value type, such a function would require receiving a type tag to know which type of concrete empty map to construct. This provides for an awkward interface to users, and it can easily be avoided by delegating the responsibility of constructing empty map constants to the compiler.
+
+`insertBSMap` and `deleteBSMap` remain fundamental operations which we must expose as builtin functions.
+
+One seemingly unnecessary addition is `unionBSMap`, as it may be implemented in Plutus Core by traversing the elements of one of the maps and calling `insertBSMap`. However, as outlined by [CPS-0013][1], `ByteStringMap` should provide an optimal implementation of the Plutus Tx `Value` type, and the operations on `Value`s depend heavily on map unions. Therefore, for performance reasons, we prefer to expose this functionality as a builtin function.
```

Comment:
I doubt that `Map`-backed `Value` is going to happen, it's way too much complexity just to get a slow version of a hypothetical built-in `Value`.